termux-exec.c

```diff
@@ -8,231 +15,508 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
-#include <elf.h>
+
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
 
 #ifndef TERMUX_BASE_DIR
-# define TERMUX_BASE_DIR "/data/data/com.termux/files"
+#define TERMUX_BASE_DIR "/data/data/com.termux/files"
```

Comment:
Do not use hardcoded  `com.termux` values here and elsewhere, it affects termux forks. To get prefix, first check `TERMUX__PREFIX` env variable, then `PREFIX`, otherwise default to package build time path `@TERMUX_BASE_DIR@`. In future, multiple rootfs support may also be added and `$PREFIX` has been deprecated in favour of `TERMUX__PREFIX` (note double dash as scope), as it conflicts with other packages like `npm`, etc. Haven't pushed change to termux-app yet.

Reply:
Thanks - I'll update!

Reply:
Changed to look at `TERMUX__PREFIX` now, with a `TERMUX_BASE_DIR` value given at compilation time as fallback.

I think `PREFIX` is too dangerous to use, as you say it can totally break when reused for other purposes.

Reply:
Thanks. But do not use literal `com.termux` value. The whole path `@TERMUX_BASE_DIR@` should be replaced during build time in the source by value set in `properties.sh`. Possibly tests should use this too.

https://github.com/termux/termux-packages/blob/26b05538619ffd93cfd52f252f34d1c025d51be7/scripts/properties.sh#L34

For inspiration, check termux-tools design by grimler.

https://github.com/termux/termux-tools/blob/master/configure.ac

https://github.com/termux/termux-tools/blob/master/Makefile.am

Also do not limit prefix length to `48`, package names can be much longer. For example `/data/user/10/net.dinglisch.android.taskerm/files` will fail with length `49`. Even `/data/data` paths can be longer. Package name alone on playstore can be of length `50`. Probably just check if it starts with `/` since technically packages can be compiled for any prefix, even `/data/local/tmp` for `adb` or `/mnt` paths for external sd.

Reply:
Yeah, good idea not to use `$PREFIX`,  will have to push `TERMUX__PREFIX` export to termux-app as well.

Reply:
> Also do not limit prefix length to 48, package names can be much longer

Right, we don't need to be so conservative here. Bumped to `TERMUX_BASE_DIR_MAX_LEN=100` - that at least should be more than enough.

Reply:
> But do not use literal com.termux value

We start by using the `TERMUX__PREFIX` env variable now. If that does not exist, we fall back to `TERMUX_BASE_DIR`.

`TERMUX_BASE_DIR` is given at compilation time, as in `make TERMUX_BASE_DIR=/a/path/to/package/prefix` (the 
It's not really hardcoded any longer, but supplied at compile time (the existing package already supplies that at build time: https://github.com/termux/termux-packages/blob/master/packages/termux-exec/build.sh#L10).

Remaining usage of `com.termux` are ok I think:
- In `Makefile`, as a default value if `make` is executed without any argument
- In `termux-exec.c`, but only inside `#ifdef UNIT_TEST`, so only used for testing. We also test with other prefixes as well in the tests.
- In `test-program.c`, which is just a utility test program that can be run manually, not part of the package

Reply:
Thanks for your thorough review and feedback @agnostic-apollo !

Reply:
Will check code in detail later, outside, but if app is installed on adoptable external sd, then its app data parent directory prefix will be `/mnt/expand/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/user/0`, which alone has length `55`, so `100` is not going to be enough, especially if multiple sub dirs under package name.

You are very welcome. :)

Reply:
> so 100 is not going to be enough

Thanks again! So what about `200`, that should be enough for everyone perhaps :)?  That allows a path such as `/mnt/expand/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/user/0/yyyyyyyyyy-yyyyyyyyyy/yyyyyyyyyy-yyyyyyyyyyyyyyyyyyyy/yyyyyyyyyyyyyyyyyyyy/yyyyyyyyyyyyyyyyyyyy-yyyyyy
yyyy/xxxxx-xxxxx/yyyyy-yyyyy/zzzzzz-zzzzz/`.

The reason I don't want an arbitrary big length is that it would be good to have a reasonable tight limit - this code is going to be injected in a lot of places, so not taking up more stack or heap space than necessary would be great, to decrease risks of interfering with the surrounding problem.

Reply:
lolz technically the only valid length is `PATH_MAX`(`4096` bytes), but that would be for the full path, not just the prefix. That is also what libraries including mine use as the standard for passing to sys calls. You could reduce it if necessary, but couple of hundred bytes won't matter, but at least log a warning if value is ignored.